---
title: "Report for `r params$name`"
author: "Esmeralda Garc√≠a Legarda"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    number_sections: false
    theme: united
params:
  name:
    label: "name"
    value: overlap
    input: file

---

```{r, setup, include=FALSE}
knitr::opts_knit$set(root.dir = "../")
```

<style>
.list-group-item.active, .list-group-item.active:focus, .list-group-item.active:hover {
    background-color: #008080;
}
.nav-pills>li.active>a, .nav-pills>li.active>a:hover, .nav-pills>li.active>a:focus {
    color: white;
    background-color: #008080;
 }
</style>

<div style="text-align: justify">

The specified libraries are loaded before the beginning of the report.

```{r, warning=FALSE, message=FALSE}
library(stringr)
library(rcartocolor)
library(ggplot2)
library(grid)
library(gridExtra)
library(cowplot)
library(dplyr)
library(DT)
```


### a) Data

The aim is to find and describe the proportion of overlapping genes. The first step was searching against the VARIABLE database by PSI-Blast software. It test if the numbers of hits of the search would be considerably different if different sequences of the same protein query is used as input. In addition, it tests how changing the *evalue* between the default value, VARIABLE, could affect the number of hits it could be retrieved, using also VARIABLE as *evalue*.

The initial hits were filtered by a minimum of percentage of identity (VARIABLE) and a minimum of alignment length (VARIABLE amino acids). Then, the number of hits which belongs to virus species was also saved. The number of total species and the number of only viral and cellular species were collected too.





